File: internal/crypto/crypto.go
// AES-256-GCM with PBKDF2-SHA256. Wire format and parameters are byte-compatible
// with chrome-extension/slack/crypto.js: base64(salt[32] || nonce[12] || ciphertext+tag).
// PBKDF2: SHA-256, 100 000 iterations, 32-byte key. If you change this, change crypto.js too.
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"io"

	"golang.org/x/crypto/pbkdf2"
)

const (
	saltLen    = 32
	nonceLen   = 12
	iterations = 100_000
	keyLen     = 32
)

func deriveKey(passphrase string, salt []byte) []byte {
	return pbkdf2.Key([]byte(passphrase), salt, iterations, keyLen, sha256.New)
}

func Encrypt(plaintext, passphrase string) (string, error) {
	salt := make([]byte, saltLen)
	if _, err := io.ReadFull(rand.Reader, salt); err != nil {
		return "", err
	}
	nonce := make([]byte, nonceLen)
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return "", err
	}
	key := deriveKey(passphrase, salt)
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}
	ct := gcm.Seal(nil, nonce, []byte(plaintext), nil)
	payload := make([]byte, 0, saltLen+nonceLen+len(ct))
	payload = append(payload, salt...)
	payload = append(payload, nonce...)
	payload = append(payload, ct...)
	return base64.StdEncoding.EncodeToString(payload), nil
}

func Decrypt(encoded, passphrase string) (string, error) {
	raw, err := base64.StdEncoding.DecodeString(encoded)
	if err != nil {
		return "", fmt.Errorf("base64 decode: %w", err)
	}
	if len(raw) < saltLen+nonceLen+1 {
		return "", fmt.Errorf("ciphertext too short")
	}
	salt := raw[:saltLen]
	nonce := raw[saltLen : saltLen+nonceLen]
	ct := raw[saltLen+nonceLen:]
	key := deriveKey(passphrase, salt)
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}
	plaintext, err := gcm.Open(nil, nonce, ct, nil)
	if err != nil {
		return "", fmt.Errorf("decrypt: %w", err)
	}
	return string(plaintext), nil
}

File: chrome-extension/slack/crypto.js
// AES-256-GCM encryption matching the Go implementation in internal/crypto/crypto.go
// Format: salt (32 bytes) || nonce (12 bytes) || ciphertext

const SALT_LEN = 32;
const NONCE_LEN = 12;

/**
 * Derive a 256-bit key from a passphrase and salt using PBKDF2.
 */
async function deriveKey(passphrase, salt, usage) {
  const encoder = new TextEncoder();
  const keyMaterial = await crypto.subtle.importKey(
    "raw",
    encoder.encode(passphrase),
    "PBKDF2",
    false,
    ["deriveKey"]
  );
  return crypto.subtle.deriveKey(
    {
      name: "PBKDF2",
      salt: salt,
      iterations: 100000,
      hash: "SHA-256",
    },
    keyMaterial,
    { name: "AES-GCM", length: 256 },
    false,
    usage
  );
}

/**
 * Encrypt plaintext with passphrase using AES-256-GCM.
 * Returns base64-encoded: salt || nonce || ciphertext
 */
async function encrypt(plaintext, passphrase) {
  const encoder = new TextEncoder();
  const salt = crypto.getRandomValues(new Uint8Array(SALT_LEN));
  const nonce = crypto.getRandomValues(new Uint8Array(NONCE_LEN));

  const key = await deriveKey(passphrase, salt, ["encrypt"]);
  const ciphertext = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: nonce },
    key,
    encoder.encode(plaintext)
  );

  const result = new Uint8Array(SALT_LEN + NONCE_LEN + ciphertext.byteLength);
  result.set(salt, 0);
  result.set(nonce, SALT_LEN);
  result.set(new Uint8Array(ciphertext), SALT_LEN + NONCE_LEN);

  return btoa(String.fromCharCode(...result));
}

/**
 * Decrypt base64-encoded ciphertext with passphrase.
 * Input format: base64(salt || nonce || ciphertext)
 */
async function decrypt(encoded, passphrase) {
  const raw = Uint8Array.from(atob(encoded), (c) => c.charCodeAt(0));

  if (raw.length < SALT_LEN + NONCE_LEN + 1) {
    throw new Error("ciphertext too short");
  }

  const salt = raw.slice(0, SALT_LEN);
  const nonce = raw.slice(SALT_LEN, SALT_LEN + NONCE_LEN);
  const ciphertext = raw.slice(SALT_LEN + NONCE_LEN);

  const key = await deriveKey(passphrase, salt, ["decrypt"]);
  const plaintext = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: nonce },
    key,
    ciphertext
  );

  return new TextDecoder().decode(plaintext);
}

// Export for use in other scripts
if (typeof globalThis !== "undefined") {
  globalThis.teamsCLICrypto = { encrypt, decrypt };
}
